HV-1973 - Bitcoin address validator

jyoshiriro opened the pull request:
Bitcoin address validator.

More translations are still missing. Only english and portuguese have been provided so far.

See https://hibernate.atlassian.net/browse/HV-1973.

marko-bekhta commented on engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java:
```diff
@@ -0,0 +1,84 @@
+		}
+
+		for ( BitcoinAddressType type : this.addressType ) {
+			Pattern pattern = Pattern.compile( type.getRegex() );
```
Let's have the patterns pre-compiled.

jyoshiriro replied:
Done! When you have a chance to take a look please :)

marko-bekhta commented on engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/BitcoinAddressValidator.java:
```diff
@@ -0,0 +1,84 @@
+
+	@Override
+	public void initialize(BitcoinAddress bitcoinAddress) {
+		BitcoinAddressType[] types = bitcoinAddress.value();
+
+		this.singleType = ( types.length == 1 );
+
+		if ( Arrays.stream( types ).anyMatch( type -> type == BitcoinAddressType.ANY ) ) {
+			this.typesDescription = BitcoinAddressType.ANY.getDescription();
+			this.addressType.addAll(
+				Arrays.stream( BitcoinAddressType.values() )
+						.filter( type -> type != BitcoinAddressType.ANY )
+						.collect( Collectors.toList() ) );
+			return;
+		}
+
+		Collections.addAll( this.addressType, types );
+		this.typesDescription = this.addressType.stream()
+								.map( BitcoinAddressType::getDescription )
+								.collect( Collectors.joining( "; " ) );
```
once the impl details are moved out of the enum here you could still keep them in the enum and maybe also have this prebuilt? e.g.:
```java
public enum BitcoinAddressConstraint {
	P2PKH("Legacy (P2PKH)", "^(1)[a-zA-HJ-NP-Z0-9]{25,61}$"),
	P2SH("Nested SegWit (P2SH)", "^(3)[a-zA-HJ-NP-Z0-9]{33}$"),
	BECH32("Native SegWit (Bech32)", "^(bc1)[a-zA-HJ-NP-Z0-9]{39,59}$"),
	P2WSH("SegWit variant of P2SH (P2WSH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{58}$"),
	P2WPKH("SegWit variant of P2PKH (P2WPKH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{38}$"),
	P2TR("Taproot (P2TR)", "^(bc1p)[a-zA-HJ-NP-Z0-9]{58}$");

	BitcoinAddressConstraint(String description, String pattern) {
		this.pattern = Pattern.compile( pattern );
	}

	private static final EnumMap<BitcoinAddress.Type, Collection<BitcoinAddressConstraint>> constraints;
	static {
		constraints = new EnumMap<>( BitcoinAddress.Type.class );
		constraints.put( BitcoinAddress.Type.ANY, Arrays.asList( BitcoinAddressConstraint.values() ) );
		constraints.put( BitcoinAddress.Type.P2PKH, Collections.singleton( P2PKH ) );
		constraints.put( BitcoinAddress.Type.P2SH, Collections.singleton( P2SH ) );
		constraints.put( BitcoinAddress.Type.BECH32, Collections.singleton( BECH32 ) );
		constraints.put( BitcoinAddress.Type.P2WSH, Collections.singleton( P2WSH ) );
		constraints.put( BitcoinAddress.Type.P2WPKH, Collections.singleton( P2WPKH ) );
		constraints.put( BitcoinAddress.Type.P2TR, Collections.singleton( P2TR ) );

	}
 ... 
}
```
and then in the initialize:
```java
public void initialize(BitcoinAddress bitcoinAddress) {
	Set<BitcoinAddressConstraint> constraints = new HashSet<>();
	for ( BitcoinAddress.Type type : bitcoinAddress.value() ) {
		constraints.addAll( BitcoinAddressConstraint.constraints.get( type ) );
	}
	this.singleType = constraints.size() == 1;
}
```

jyoshiriro replied:
I did almost that... When you have a chance to take a look please :)

marko-bekhta commented on engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java:
```diff
@@ -0,0 +1,38 @@
+
+public enum BitcoinAddressType {
+	ANY("Bitcoin", null),
+	P2PKH("Legacy (P2PKH)", "^(1)[a-zA-HJ-NP-Z0-9]{25,61}$"),
+	P2SH("Nested SegWit (P2SH)", "^(3)[a-zA-HJ-NP-Z0-9]{33}$"),
+	BECH32("Native SegWit (Bech32)", "^(bc1)[a-zA-HJ-NP-Z0-9]{39,59}$"),
+	P2WSH("SegWit variant of P2SH (P2WSH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{58}$"),
+	P2WPKH("SegWit variant of P2PKH (P2WPKH)", "^(bc1q)[a-zA-HJ-NP-Z0-9]{38}$"),
+	P2TR("Taproot (P2TR)", "^(bc1p)[a-zA-HJ-NP-Z0-9]{58}$");
```
BTW I see that this description is used in the validation message. But as they use regular words they may need to be translated as well... So I'd suggest exploring an option of placing the description in the translation file as well and then let EL do the string interpolation.

jyoshiriro replied:
Done. When you have a chance to take a look please :)

marko-bekhta reviewed (commented):
Hey,

thanks for submitting the pull request with the new constraint. This looks interesting. 
I've added some suggestions inline. Please also take a look at this post https://in.relation.to/2018/01/04/adding-new-constraint-to-engine/ to see about the additional parts that should be added for a constraint.

jyoshiriro commented on engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java:
```diff
@@ -0,0 +1,38 @@
+ */
+package org.hibernate.validator.constraints;
+
+public enum BitcoinAddressType {
```
I did almost that... When you have a chance to take a look please :)

jyoshiriro commented on engine/src/main/java/org/hibernate/validator/constraints/BitcoinAddressType.java:
```diff
@@ -0,0 +1,38 @@
+	@Override
+	public String toString() {
+		return this.description;
+	}
```
I did almost that... When you have a chance to take a look please :)

jyoshiriro commented:
> Hey,
> 
> thanks for submitting the pull request with the new constraint. This looks interesting. I've added some suggestions inline. Please also take a look at this post https://in.relation.to/2018/01/04/adding-new-constraint-to-engine/ to see about the additional parts that should be added for a constraint.

I believe I fixed everything that was missing :)

marko-bekhta commented:
Thanks! I've squashed the changes in this patch and removed the dependency on the resource bundle from the validator implementation. I'm going to close this PR in favour of  https://github.com/hibernate/hibernate-validator/pull/1387 
Thanks for all the work you did here!
